grovel/grovel.lisp

```diff
@@ -179,68 +195,140 @@ int main(int argc, char**argv) {
+          (dolist (form header-forms)
+            (process-grovel-form out form))
+          (write-string *prologue* out)
+          (dolist (form body-forms)
+            (process-grovel-form out form))
+          (write-string *postscript* out)
+          c-file)))))
 
 (defun tmp-lisp-file-name (defaults)
   (make-pathname :name (strcat (pathname-name defaults) ".grovel-tmp")
                  :type "lisp" :defaults defaults))
 
-
+
+(defun read-grovel-file (input-file)
+  (flet ((read-forms (s)
+           (do ((forms ()) (form (read s nil nil) (read s nil nil)))
+               ((null form) (nreverse forms))
+             (labels
+                 ((process-form (f)
+                    (case (form-kind f)
+                      (flag (warn "Groveler clause FLAG is deprecated, use CC-FLAGS instead.")))
+                    (case (form-kind f)
+                      (in-package
+                       (setf *package* (find-package (second f)))
+                       (push f forms))
+                      (progn
+                        ;; flatten progn forms
+                        (mapc #'process-form (rest f)))
+                      (t (push f forms)))))
+               (process-form form)))))
+    (with-open-file (in input-file :direction :input)
+      (with-cached-reader-conditionals (read-forms in)))))
+
+(defun touch-file (pathname)
+  (with-open-file (stream pathname :direction :output
+                          :if-exists :append
+                          :if-does-not-exist :create)))
+
+(defun feature-specific-cache-dir (cache-dir feature-expressions)
+  (let ((feature-expressions (copy-seq feature-expressions)))
+    (when cache-dir
+      (ensure-directory-pathname
+       (subpathname cache-dir (gen-feature-hash feature-expressions))))))
+
+(defun feature-specific-cache-file (file-name cache-dir feature-expressions)
+  (let ((fs-cache-dir (feature-specific-cache-dir
+                       cache-dir feature-expressions)))
+    (when fs-cache-dir
+      (subpathname fs-cache-dir (pathname-name file-name)
+                   :type (pathname-type file-name)))))
+
+(defun ensure-fresh-dir (abs-path)
+  (assert (uiop:absolute-pathname-p abs-path))
+  (assert (uiop:directory-pathname-p abs-path))
+  (when (uiop:directory-exists-p abs-path)
+    (uiop:delete-directory-tree abs-path :validate t))
+  (ensure-directories-exist abs-path))
+
+(defun copy-local-includes-to-cache (in-dir)
+  (assert (boundp '*local-includes*))
+  (assert *local-includes*)
+  (let ((local-include-dir
+         (uiop:ensure-directory-pathname
+          (subpathname (uiop:pathname-directory-pathname in-dir)
+                       "local-includes"))))
+    (ensure-fresh-dir local-include-dir)
+    (loop :for (src-file . local-name) :in *local-includes*
+       :for dest-file := (subpathname local-include-dir local-name)
+       :do (alexandria:copy-file src-file dest-file))
+    (format nil "-I~A" (truename local-include-dir))))
```

Comment:
This function is part of the `include-local` mechanism. When we discussed this feature in the mailing list, stelian said:

> I think it's better for the groveler to add the source directory to the include path, so that the distinction doesn't matter. with grovel.c being
>
> #include <foo.h>
>
> `cc -I/path/to/sources/ grovel.c` will then search for foo.c first in /path/to/sources/ and then in the default system include path.

To enable this we create a directory in the `.cache/common-lisp/etc/` cache directory and copy all the `local-include` header files into it. We later use `-I` to add this directory to the search path as suggested (which works perfectly). If you recall, in `push-local-include` we create a unique name for each `local-include` file so we know they won't clash. 

We also use `ensure-fresh-dir` to make sure any header files from previous builds are removed.